lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -2593,93 +2677,116 @@ class ConvertAtenMaxPool2dOp : public OpConversionPattern<AtenMaxPool2dOp> {
-    // will get the lowered version of the operands which is harder to pattern
-    // match.
-    SmallVector<int64_t, 2> strideInts;
-    if (!matchPattern(op.stride(), m_TorchConstantIntList(strideInts)))
-      return rewriter.notifyMatchFailure(op,
-                                         "only support constant int strides");
-    SmallVector<int64_t, 2> dilationInts;
-    if (!matchPattern(op.dilation(), m_TorchConstantIntList(dilationInts)))
-      return rewriter.notifyMatchFailure(op,
-                                         "only support constant int dilations");
-    SmallVector<int64_t, 2> paddingInts;
-    if (!matchPattern(op.padding(), m_TorchConstantIntList(paddingInts)))
-      return rewriter.notifyMatchFailure(op,
-                                         "only support constant int paddings");
-    SmallVector<int64_t, 2> kernelSizeInts;
-    if (!matchPattern(op.kernel_size(), m_TorchConstantIntList(kernelSizeInts)))
-      return rewriter.notifyMatchFailure(op, "only support kernel size ints");
-
-    Value falseValue = rewriter.create<arith::ConstantOp>(
-        loc, IntegerAttr::get(rewriter.getIntegerType(1), 0));
-    Value ceilModeFalse = rewriter.create<arith::CmpIOp>(
-        loc, arith::CmpIPredicate::eq, ceilMode, falseValue);
-    rewriter.create<AssertOp>(
-        loc, ceilModeFalse,
-        rewriter.getStringAttr("only ceil_mode false is supported"));
+    Value self = adaptor.self();
+    Value maxPool2d;
+    if (failed(
+            computeMaxPool2d<AtenMaxPool2dOp>(op, rewriter, self, maxPool2d)))
+      return rewriter.notifyMatchFailure(op, "unable to compute maxpool2d");
+    Type newResultType = getTypeConverter()->convertType(op.getType());
+    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, newResultType, maxPool2d);
+    return success();
+  }
+};
+} // namespace
 
-    SmallVector<int64_t, 4> paddingIncludingNC = {0, 0};
-    paddingIncludingNC.insert(paddingIncludingNC.end(), paddingInts.begin(),
-                              paddingInts.end());
-    Value paddedInput = getPaddedTensor(op, rewriter, self, paddingIncludingNC);
+namespace {
+// Returns the result of maxpool2d over the input tensor. And the corresponding
+// indices of the input tensor for the values of the result tensor.
+//
+// The result of the maxpool2d operation is calculated using the helper function
+// written above. For finding the indices, we follow the below method:
+//
+// Let's say the input tensor is a 4-d tensor. The maxpool2d and indices will
+// also be a 4-d tensor. Then:
+//  for i in input.size[0]:
+//    for j in input.size[1]:
+//      for k in input.size[2]:
+//        for l in input.size[3]:
+//          for p in maxpool2d.size[1]:
+//            for q in maxpool2d.size[2]:
+//              for r in maxpool2d.size[3]:
+//                if input[i, j, k, l] == maxpool2d[i, p, q, r]:
+//                  indices[i, p, q, r] = (k * input.size[3] + l)
```

Comment:
I don't think this algorithm is correct. Lets say the input tensor has two windows `A` and `B`. If a value `x` in `B` happens to equal the maximum of `A`, then the index stored for window `A` could end up being the index of `x`, which is in a different window. I'm also worried a bit about the performance since we are iterating over almost the entire `maxpool2d` tensor for every value in the input tensor.

One way of doing this would be to restrict the search to the window you're currently in. From the values of `k` and `l`, doing some arithmetic gymnastics, one should be able to determine the window that that point is in, allowing you to compare it to the max value for that window in the `maxpool2d` tensor. This removes the need to iterate over the `maxpool2d` tensor, and it ensures that we are only getting indices from the window we are currently examining.

The way I mentioned above is also close to how one could calculate `maxpool2d` in the first place. It seems that to get the indices we are essentially implementing a variation on `maxpool2d` by hand, which is also a bit concerning, since now `maxpool2d_indices` will take at least twice as long as regular `maxpool2d`. Ideally the extraction of the indices should be performed at the same time as the pooling, but this would require changes to linalg. For now I think it is okay to have `O(2 * maxpool2d)` performance.

Reply:
Hi @ramiro050, I have a doubt. What if the maximum value appears twice in a window? Which of the index should we choose? Also, do you think it would be easier to implement this in TMTensor? I am not sure where to implement this.

Reply:
Hi @ramiro050 @cathyzhyi, I have thought of the following algorithm for this:

The result of the max pool operation is `maxPool2d`, and its shape is (N, C, Hout, Wout), and the kernel size is (kH, kW). Then
```
for i in range(N):
    for j in range(C):
        for m in range(Hout):
            for n in range(Wout):
                for p in range(kH):
                    for r in range(kW):
                        if input[i, j, m + p, n + r] == maxPool2d[i, j, m, n]:
                            indices[i, j, m, n] = (m + p)*kW + (n + r)
```
This algorithm handles the padding implicitly but does not handle the strides and dilation. I am thinking of a way to get this work for the cases where dilation and strides do not equal one, which is the default value. If you have any suggestions, please do comment. I am implementing this one until I find some way for the non-default stride and dilation to get work.

Reply:
I implemented this algorithm and tested it just now. It works for the default dilation, default stride, and unpadded input case. I will check why it's failing for the padding case for which it should have worked.

Reply:
I think to consider stride and dilation you just need to adjust the input index:
```
		index_H = m*stride[0] + p* dilation[0]
		index_W = n*stride[1] + r* dilation[1]
                if input[i, j, index_H, index_W] == maxPool2d[i, j, m, n]:
                      indices[i, j, m, n] = index_H*kW + index_W
```

Reply:
Able to get this working for all non-default params.